python-lib/dku_webapp/utils.py

```diff
@@ -0,0 +1,182 @@
+import dataiku
+import pandas as pd
+import numpy as np
+import logging
+from flask import request
+import simplejson
+import traceback
+from dku_model_accessor import get_model_handler, ModelAccessor
+from dku_model_fairness_report import ModelFairnessMetricReport, ModelFairnessMetric
+from dku_model_fairness_report.constants import DkuFairnessConstants
+from dku_webapp.constants import DkuWebappConstants
+
+
+
+def get_histogram_data(y_true, y_pred, y_pred_proba, advantageous_outcome, sensitive_feature_values):
+    df = pd.DataFrame()
+    # the following strings are used only here, too lazy to turn them into constant variables
+    df['bin_index'] = np.digitize(y_pred_proba, np.arange(0, 1, step=0.1))  # 10 bins
+    df['prediction_result_type'] = get_prediction_result_type(y_true, y_pred, advantageous_outcome)
+    df['sensitive_feature'] = sensitive_feature_values
+
+    cast_to_int = False
+    try: # check whether or not the column can be casted to int
+        if np.array_equal(df['sensitive_feature'], df['sensitive_feature'].astype(int)):
+            cast_to_int = True
+    except:
+        pass
+
+    histogram_dict = {}
+    for v in df['sensitive_feature'].unique():
+
+        df2 = df[df['sensitive_feature'] == v]
+        dfx = np.round(100 * df2.groupby(['prediction_result_type', 'bin_index']).size() / len(df2), 3)
+        series_final = dfx.unstack().fillna(0).stack()
+
+        computed_df = pd.DataFrame(series_final, columns=['bin_value_new'])
+        arrays = [['predicted_0_true_0', 'predicted_0_true_1', 'predicted_1_true_0', 'predicted_1_true_1'], np.arange(1, 11)]
+        reference_df = pd.DataFrame(index=pd.MultiIndex.from_product(arrays, names=('prediction_result_type', 'bin_index')))
+        reference_df['bin_value'] = 0
+        result_df = pd.concat([reference_df, computed_df], axis=1, sort=False)
+        # replace missing value by 0
+        result_df['bin_value_final'] = result_df['bin_value_new'].fillna(result_df['bin_value'])
+
+        dct = {}
+        for i in result_df.index.levels[0]:
+            dct[i] = result_df.xs(i, level='prediction_result_type')['bin_value_final'].values.tolist()
+        if cast_to_int:
+            histogram_dict[int(v)] = dct
+        else:
+            histogram_dict[v] = dct
+
+
+    return histogram_dict
+
+
+def convert_numpy_int64_to_int(o):
+    if isinstance(o, np.int64):
+        return int(o)
+    raise TypeError
+
+
+def get_prediction_result_type(y_true, y_pred, advantageous_outcome):
+    df = pd.DataFrame()
+    # the following strings are used only here, too lazy to turn them into constant variables
```

Comment:
To me it is not a surprise to have the value hardcoded

Reply:
👌